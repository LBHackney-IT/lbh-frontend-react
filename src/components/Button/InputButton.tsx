import classNames from "classnames";
import { nullValuesAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";

import { Attributes, DataAttributes } from "../../helpers/Attributes";
import { debounce, makeDebounceContext } from "../../helpers/debounce";
import "lbh-frontend/lbh/components/lbh-button/_button.scss";

/**
 * The proptypes for {@link InputButton}.
 *
 * This also supports all `aria-*` and `data-*` attributes.
 *
 * @noInheritDoc
 */
export interface InputButtonProps extends React.AriaAttributes, DataAttributes {
  id?: string | null;
  className?: string | null;
  /**
   * Use for the main call to action on your service's start page. Adjusts the
   * styling of the button but, unlike {@link Button}, does not add an icon due
   * to this using an `<input>` internally.
   */
  isStartButton?: boolean | null;
  /**
   * Debounces clicks within a second of each other preventing multiple form
   * submissions.
   */
  preventDoubleClick?: boolean | null;
  name?: string | null;
  type: React.ButtonHTMLAttributes<HTMLButtonElement>["type"];
  value?: string | null;
  disabled?: boolean | null;
}

/**
 * A component for creating an input element, styled as a button. It can be
 * used to help users carry out actions such as submitting forms.
 *
 * @noInheritDoc
 */
export class InputButton extends React.Component<InputButtonProps> {
  static propTypes: PropTypes.ValidationMap<InputButtonProps> = {
    id: PropTypes.string,
    className: PropTypes.string,
    isStartButton: PropTypes.bool,
    preventDoubleClick: PropTypes.bool,
    name: PropTypes.string,
    type: PropTypes.oneOf<"submit" | "button" | "reset">([
      "submit",
      "button",
      "reset"
    ]).isRequired,
    value: PropTypes.string,
    disabled: PropTypes.bool
  };

  private debounceContext = makeDebounceContext();

  private handleClick(event: React.MouseEvent<HTMLInputElement>): void {
    const { disabled } = this.props;

    if (disabled) {
      event.preventDefault();
      return;
    }

    if (debounce(event, this.debounceContext)) {
      return;
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const {
      id,
      className,
      isStartButton,
      preventDoubleClick,
      name,
      type,
      value,
      disabled
    } = nullValuesAsUndefined(this.props);

    const extraAttributes = Attributes.ariaAndData(this.props);

    return (
      <input
        id={id}
        className={classNames("govuk-button lbh-button", className, {
          "govuk-button--start": isStartButton
        })}
        name={name}
        value={value}
        type={type}
        disabled={disabled}
        onClick={this.handleClick.bind(this)}
        {...extraAttributes}
        aria-disabled={disabled}
        data-prevent-double-click={preventDoubleClick}
      />
    );
  }
}
