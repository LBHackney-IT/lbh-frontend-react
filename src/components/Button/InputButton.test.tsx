import React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

import { InputButton } from "./InputButton";

it("renders correctly with all props", () => {
  const component = create(
    <InputButton
      id="1234"
      className="class1 class2"
      isStartButton
      preventDoubleClick
      name="Testname"
      type="submit"
      value="Testvalue"
      disabled
    />
  );

  expect(component).toMatchInlineSnapshot(`
    <input
      aria-disabled={true}
      className="govuk-button lbh-button class1 class2 govuk-button--start"
      data-prevent-double-click={true}
      disabled={true}
      id="1234"
      name="Testname"
      onClick={[Function]}
      type="submit"
      value="Testvalue"
    />
  `);
});

it("renders correctly without optional props", () => {
  const component = create(<InputButton type="submit" />);

  expect(component).toMatchInlineSnapshot(`
    <input
      className="govuk-button lbh-button"
      onClick={[Function]}
      type="submit"
    />
  `);
});

it("prevents a second click within a second from executing the button a second time when preventDoubleClick is true", () => {
  const wrapper = mount(
    <InputButton preventDoubleClick type="submit">
      Testchildren
    </InputButton>
  );

  const input = wrapper.find("input");
  const inputOnClick = input.props().onClick;

  expect(inputOnClick).toBeDefined();

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: input.getDOMNode(),
    preventDefault: mockPreventDefault
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  inputOnClick!.call(input, mockClickEvent);

  expect(mockPreventDefault).not.toHaveBeenCalled();

  jest.advanceTimersByTime(999);

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  inputOnClick!.call(input, mockClickEvent);

  expect(mockPreventDefault).toHaveBeenCalledTimes(1);
});

it("allows a second click after a second to execute the button a second time when preventDoubleClick is true", () => {
  const wrapper = mount(
    <InputButton preventDoubleClick type="submit">
      Testchildren
    </InputButton>
  );

  const input = wrapper.find("input");
  const inputOnClick = input.props().onClick;

  expect(inputOnClick).toBeDefined();

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: input.getDOMNode(),
    preventDefault: mockPreventDefault
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  inputOnClick!.call(input, mockClickEvent);

  expect(mockPreventDefault).not.toHaveBeenCalled();

  jest.advanceTimersByTime(1000);

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  inputOnClick!.call(input, mockClickEvent);

  expect(mockPreventDefault).not.toHaveBeenCalled();
});

it("prevents a click from executing the button when disabled is true", () => {
  const wrapper = mount(
    <InputButton preventDoubleClick type="submit" disabled>
      Testchildren
    </InputButton>
  );

  const input = wrapper.find("input");
  const inputOnClick = input.props().onClick;

  expect(inputOnClick).toBeDefined();

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: input.getDOMNode(),
    preventDefault: mockPreventDefault
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  inputOnClick!.call(input, mockClickEvent);

  expect(mockPreventDefault).toHaveBeenCalledTimes(1);
});
