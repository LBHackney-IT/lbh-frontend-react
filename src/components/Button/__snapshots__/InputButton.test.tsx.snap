// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<input
  aria-disabled={true}
  className="govuk-button lbh-button class1 class2 govuk-button--start"
  data-prevent-double-click={true}
  disabled={true}
  id="1234"
  name="Testname"
  onClick={[Function]}
  type="submit"
  value="Testvalue"
/>
`;

exports[`renders correctly without optional props 1`] = `
<input
  className="govuk-button lbh-button"
  onClick={[Function]}
  type="submit"
/>
`;
